src/args.rs

```diff
@@ -0,0 +1,575 @@
+	de::{SeqAccess, Visitor},
+	ser::SerializeSeq,
+	*,
+};
+use thiserror::Error;
+
+#[cfg(target_os = "linux")]
+use crate::params::FileSandboxMode;
+use crate::params::{CpuCount, EnvVars, GuestMemorySize, Output, Params};
+
+/// Used by clap to derive CLI parameters for Uhyve.
+#[derive(Parser, Debug)]
+#[clap(version, author, about)]
+pub struct Args {
+	#[clap(flatten, next_help_heading = "Uhyve OPTIONS")]
+	pub uhyve_args: UhyveArgs,
+
+	#[clap(flatten, next_help_heading = "Memory OPTIONS")]
+	pub memory_args: MemoryArgs,
+
+	#[clap(flatten, next_help_heading = "Cpu OPTIONS")]
+	pub cpu_args: CpuArgs,
+
+	#[clap(flatten, next_help_heading = "Guest OPTIONS")]
+	pub guest_args: GuestArgs,
+}
+
+/// This is the config that defines a set of parameters for a given Hermit machine image.
+///
+/// Note that the field names are relevant for the TOML's tables.
+///
+/// TODO: Figure out how to adapt file isolation.
+#[derive(Debug, Deserialize)]
+pub struct UhyveGuestConfig {
+	pub(crate) memory: MemoryArgs,
+	pub cpu: CpuArgs,
+	pub guest: GuestArgs,
+}
+
+/// Arguments for Uhyve runtime-related configurations.
+#[derive(Parser, Debug)]
+pub struct UhyveArgs {
+	/// Kernel output redirection.
+	///
+	/// None discards all output, Omit for stdout
+	#[clap(short, long, value_name = "FILE")]
+	pub(super) output: Option<String>,
+
+	/// Display statistics after the execution
+	#[clap(long)]
+	pub stats: bool,
+
+	/// Paths that the kernel should be able to view, read or write.
+	///
+	/// Desired paths must be explicitly defined after a colon.
+	///
+	/// Example: --file-mapping host_dir:guest_dir --file-mapping file.txt:guest_file.txt
+	#[clap(long)]
+	file_mapping: Vec<String>,
+
+	/// The path that should be used for temporary directories storing unmapped files.
+	///
+	/// This is useful for manually created tmpfs filesystems and for selecting
+	/// directories not managed by a temporary file cleaner, which can remove open files
+	/// manually. In most cases, mapping the guest path /root/ instead should be sufficient.
+	///
+	/// Defaults to /tmp.
+	#[clap(long)]
+	tempdir: Option<String>,
+
+	/// File isolation (none, normal, strict)
+	///
+	/// - 'none' disables all file isolation features
+	///
+	/// - 'normal' enables all file isolation features supported on the host system
+	///
+	/// - 'strict' enforces the highest amount of file isolation possible, fails on systems
+	///   that do not support them (e.g. a Linux kernel without Landlock support)
+	///
+	/// [default: normal]
+	#[clap(long)]
+	#[cfg(target_os = "linux")]
+	file_isolation: Option<String>,
+
+	/// GDB server port
+	///
+	/// Starts a GDB server on the provided port and waits for a connection.
+	#[clap(short = 's', long, env = "HERMIT_GDB_PORT")]
+	#[cfg(target_os = "linux")]
+	gdb_port: Option<u16>,
+}
+
+#[derive(Default, Parser, Debug, Serialize, Deserialize)]
+pub struct MemoryArgs {
+	/// Guest RAM size
+	#[clap(short = 'm', long, default_value_t, env = "HERMIT_MEMORY_SIZE")]
```

Comment:
I'm not that much of a fan of moving the clap parts out of `src/bin/uhyve.rs`. 
IMHO, only `UhyveGuestConfig` and TOML parsing to `UhyveGuestConfig` (using private helper structs if necessary) should be part of _uhyvelib_.

Reply:
that is fair enough :) will fix that

Reply:
I tried to implement this fyi but the byproduct is a lot of `pub` fields, I might rename the file `src/args.rs` later down the line.